planner/core/rule_column_pruning.go

```diff
@@ -348,8 +349,13 @@ func (ds *DataSource) PruneColumns(parentUsedCols []*expression.Column, opt *log
 		var handleCol *expression.Column
 		var handleColInfo *model.ColumnInfo
 		// case 1: tiflash
-		if ds.tableInfo.TiFlashReplica != nil {
+		_, isolationReadEnginesHasTiFlash := ds.ctx.GetSessionVars().GetIsolationReadEngines()[kv.TiFlash]
+		if ds.tableInfo.TiFlashReplica != nil && isolationReadEnginesHasTiFlash {
 			handleCol, handleColInfo = preferNotNullColumnFromTable(ds)
+			// The selected columns need to be added to `colsRequiringFullLen`.
```

Comment:
the tiflash store shouldn't consider index? 
is `isolationReadEnginesHasTiFlash` enough?

Reply:
I am afraid `isolationReadEnginesHasTiFlash ` is not enough.
It is possible that if both read engine and replica are satisfied, tikv is still used.

Reply:
potential that we prefer a not null column while missing the index-single-scan candidate in TiKV storage read？